artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java

```diff
@@ -319,6 +321,8 @@
     */
    private final Object directDeliveryGuard = new Object();
 
+   private final ConcurrentHashSet<String> lingerSessionIds = new ConcurrentHashSet<>();
```

Comment:
How are sure we dont leak here if sessions are closed

Reply:
Closing ServerSession will roll back transaction. And I add synchronized transaction block to avoid adding session ids when transaction is rollbacking/committing. This is what ServerSessionImpl::addLingerConsumer does(synchronizing access by this and afterCommit/afterRollback).

Reply:
Understood